### DISCUSSION TITLE
I wanna make a Waveform

### DISCUSSION THREAD

**Discussion by SleepyzDebugger** [godotengine/godot-proposals#10982] [Scripting]
I want to make a waveform but there doesn't seem to be a function to do that, the many ways i see seem outdated and also don't work at all in the version of godot, i want to have more things for .wav but there doesn't seem to be much functionality for importing a wav from fileaccess and having it work to generate a waveform, i feel like it would make sense for wav to be able to become a waveform since it does sound like waveform for short, like AudioStreamWAV.GenerateTexture or something, since AudioStreamWAV could use some more functionality other than just being .new(), so it could be used in more cases like ogg, but in a good way

**Comment by Meorge** [comment]
Isn't this what [AudioStreamGenerator](https://docs.godotengine.org/en/stable/classes/class_audiostreamgenerator.html) is for?

**Reply by SleepyzDebugger** [reply]
i tried it but, doesn't seem to wanna do what i want it to do, either that, or i just don't understand how to use it

**Reply by Meorge** [reply]
Can you be more specific about what you're looking for when you say "make a waveform"? I understood it to mean that you wanted to generate raw audio data, then convert it into sounds that you can hear. That's what AudioStreamGenerator is for

**Reply by SleepyzDebugger** [reply]
not raw audio data, just something that is like a visual waveform that is seen basically anywhere, you know seeing what part of a song your at via a waveform, i want to generate one when the user loads a song of their own or so
![image](https://github.com/user-attachments/assets/12ff3e96-e73c-44cb-b0f2-4b2130a17fe2)

**Reply by SleepyzDebugger** [reply]
so i can put that in a spot where'd i need it, and not add way more code to my audiostream node than i need to

**Reply by Meorge** [reply]
Ah, okay, I understand what you mean now.

Would you like this functionality for an editor tool/addon/etc, or for a game? I could see there being issues with a waveform image creator intended for games: different games will want their waveforms to look different, depending on the color palette, resolution, general UI direction, et cetera. So a sufficient solution for creating waveform images would need to be able to style the waveform in a variety of different ways, and might not be worth the amount of effort that would go into maintaining it, versus just having people construct their own waveform images to match the look they want.

For editor plugins, though, I think it'd make sense to have a function that outputs waveform images matching how the editor itself renders them, so that plugins feel more native within Godot.

**Reply by SleepyzDebugger** [reply]
yeah, probably a plugin, i was needing one for this functionality, to show the song progress you are at, where it would generate one once you put in a audio file, currently my game only supports .ogg loading by how loading external audio works for some reason, which idk if that can be even turned into a waveform
![image](https://github.com/user-attachments/assets/72fc92cf-3a90-42a6-98d9-276b59d16f38)

**Reply by SleepyzDebugger** [reply]
since so far this code of mine, can only access oggvorbis, i wish there were like a class as well that can just load any audio type, as a stream for a audiostreamplayer, so that the user won't be limited to ogg's, when trying to load a song from their file system
![image](https://github.com/user-attachments/assets/63ec5ac0-3332-47a6-96e7-d7a8dea8fc8d)

**Reply by SleepyzDebugger** [reply]
the conducter and h_slider here are fine with any audio, that is just to calibrate the rest of my editor like thing, however by how audiostreams work somehow, why can you load a ogg vorbis but not other types? this was one road block i encountered when just trying for songs

**Reply by Meorge** [reply]
I imagine you'd want to read the file extension from the song's filepath, and use it to determine which subclass of `AudioStream` to use (e.g., if it ends with `.mp3`, then use `AudioStreamMP3`, and if it ends with `.ogg`, then use `AudioStreamOggVorbis`)

**Reply by SleepyzDebugger** [reply]
yup, pretty much, but i haven't found a correct way to load a audiostreammp3 from searching around and testing myself

**Reply by Meorge** [reply]
Try https://docs.godotengine.org/en/stable/classes/class_audiostreammp3.html#class-audiostreammp3-property-data

**Reply by SleepyzDebugger** [reply]
i forgot how one'd read a file extension of a file and do so accordingly

**Reply by SleepyzDebugger** [reply]
well, i tried the mp3 thing, did not seem to work, even by itself

**Reply by SleepyzDebugger** [reply]
since my game sorta just relies on having only one active audio stream player, since it has the beats per second and so and does things based on those, so i'd basically would need a way like ogg to load other types

**Reply by Meorge** [reply]
I think this is something you'd be best off looking for help with on the Godot Forums, Discord server, etc rather than here in the proposals forum

**Reply by SleepyzDebugger** [reply]
Well, i did try that, radio silence there

**Reply by mk56-spn** [reply]
> Well, i did try that, radio silence there

A pr that just got merged recently will likely aid you in generating it in a convenient way #86539

**Reply by SleepyzDebugger** [reply]
i don't know what you mean by that

**Reply by Meorge** [reply]
They're referring to this PR: https://github.com/godotengine/godot/pull/86539

Its `mix_audio` method on AudioStreamPlayback sounds like it will allow you to get the audio playback data (what you use to generate a waveform) from code

**Reply by Akosmo** [reply]
Skimmed a bit. There is probably a way to generate the waveform of the entire song like that by accessing the audio bytes and displaying in some way, since audio samples should be there (and that's what you're looking for). But if you wouldn't mind displaying audio samples in real-time to make a waveform, use `AudioStreamCapture`. I plan on making a demo to demonstrate how, but reading the documentation should hopefully be clear enough.
